node/src/components/storage.rs

```diff
@@ -1616,87 +1583,226 @@ fn initialize_block_body_v1_db(
+        {
+            Some(deploy_hashes_with_proof) => deploy_hashes_with_proof,
+            None => return Ok(None),
+        };
+    let transfer_hashes_with_proof: MerkleLinkedListNode<Vec<DeployHash>> =
+        match get_merkle_linked_list_node(
+            tx,
+            block_body_v2_db,
+            transfer_hashes_db,
+            deploy_hashes_with_proof.merkle_proof_of_rest(),
+        )? {
+            Some(transfer_hashes_with_proof) => transfer_hashes_with_proof,
+            None => return Ok(None),
         };
-    tx.del(*value_database, &key_to_part_db, None)?;
-    tx.del(*block_body_v2_db, &key, None)?;
-    Ok(merkle_proof_of_rest)
+    let proposer_with_proof: MerkleLinkedListNode<PublicKey> = match get_merkle_linked_list_node(
+        tx,
+        block_body_v2_db,
+        proposer_db,
+        transfer_hashes_with_proof.merkle_proof_of_rest(),
+    )? {
+        Some(proposer_with_proof) => {
+            debug_assert_eq!(*proposer_with_proof.merkle_proof_of_rest(), hash::SENTINEL1);
+            proposer_with_proof
+        }
+        None => return Ok(None),
+    };
+    let block_body = BlockBody::new(
+        proposer_with_proof.take_value(),
+        deploy_hashes_with_proof.take_value(),
+        transfer_hashes_with_proof.take_value(),
+    );
+
+    Ok(Some(block_body))
 }
 
-/// Writes a single merklized block body in a separate transaction to storage.
-fn delete_single_block_body_v2(
-    tx: &mut RwTransaction,
+fn garbage_collect_block_body_v2_db(
+    txn: &mut RwTransaction,
     block_body_v2_db: &Database,
     deploy_hashes_db: &Database,
     transfer_hashes_db: &Database,
-    digest_key_to_delete: Digest,
+    proposer_db: &Database,
+    block_body_hash_to_header_map: &BTreeMap<Digest, BlockHeader>,
 ) -> Result<(), LmdbExtError> {
-    info!(?digest_key_to_delete, "deleting deploy hashes");
-    let digest_key_to_delete = delete_merkle_block_body_part(
-        tx,
-        block_body_v2_db,
-        deploy_hashes_db,
-        digest_key_to_delete,
-    )?;
-    info!(?digest_key_to_delete, "deleting transfer hashes");
-    delete_merkle_block_body_part(
-        tx,
-        block_body_v2_db,
-        transfer_hashes_db,
-        digest_key_to_delete,
-    )?;
-    // Don't delete proposer since it might be shared
+    // This will store all the keys that are reachable from a block header, in all the parts
+    // databases (we're basically doing a mark-and-sweep below).
+    let mut live_digests = HashSet::new();
+
+    for (body_hash, header) in block_body_hash_to_header_map {
+        if header.hashing_algorithm_version() != HashingAlgorithmVersion::V2 {
+            // We're only interested in body hashes for V2 blocks here
+            continue;
+        }
+        let mut current_digest = *body_hash;
+        loop {
+            if current_digest == hash::SENTINEL1 || live_digests.contains(&current_digest) {
+                break;
+            }
```

Comment:
Maybe `while` is cleaner than `loop … break`:

```suggestion
        while current_digest != hash::SENTINEL1 && !live_digests.contains(&current_digest) {
```

Reply:
Good idea. The condition for breaking was in the middle of the loop before, so it wouldn't make sense as a while loop, but now there are no obstacles to do that :+1: